tests/unit/h/accounts/schemas_test.py

```diff
@@ -374,43 +374,140 @@ def models(self, patch):
         return models
 
 
-@pytest.mark.usefixtures("user_password_service")
-class TestPasswordChangeSchema:
-    def test_it_is_invalid_if_passwords_dont_match(self, pyramid_csrf_request):
-        user = Mock()
-        pyramid_csrf_request.user = user
-        schema = schemas.PasswordChangeSchema().bind(request=pyramid_csrf_request)
+class TestPasswordAddSchema:
+    def test_valid(self, schema, valid_data):
+        appstruct = schema.deserialize(valid_data)
 
-        with pytest.raises(colander.Invalid) as exc:
-            schema.deserialize(
+        assert appstruct == {"csrf_token": None, **valid_data}
+
+    def test_invalid_csrf_token(self, schema, valid_data, pyramid_request):
+        del pyramid_request.headers["X-CSRF-Token"]
+
+        with pytest.raises(BadCSRFToken):
+            schema.deserialize(valid_data)
+
+    @pytest.mark.parametrize(
+        "data,expected_error_dict",
+        [
+            pytest.param(
+                {"new_password": "pass", "new_password_confirm": "pass"},
+                {"new_password": "Must be 8 characters or more."},
+                id="password_too_short",
+            ),
+            pytest.param(
                 {
-                    "new_password": "foo-bar-baz",
-                    "new_password_confirm": "foo-bar-buzz",
-                    "password": "flibble",
-                }
-            )
+                    "new_password": "valid_password",
+                    "new_password_confirm": "doesnt_match",
+                },
+                {"new_password_confirm": "The passwords must match."},
+                id="passwords_dont_match",
+            ),
+            pytest.param(
+                {"new_password": "foo", "new_password_confirm": "bar"},
+                {
+                    "new_password": "Must be 8 characters or more.",
+                },
+                id="password_too_short_and_passwords_dont_match",
+            ),
+        ],
+    )
+    def test_invalid(self, schema, data, expected_error_dict):
+        with pytest.raises(colander.Invalid) as exc_info:
+            schema.deserialize(data)
 
-        assert "new_password_confirm" in exc.value.asdict()
+        assert exc_info.value.asdict() == expected_error_dict
 
-    def test_it_is_invalid_if_current_password_is_wrong(
-        self, pyramid_csrf_request, user_password_service
-    ):
-        user = Mock()
-        pyramid_csrf_request.user = user
-        schema = schemas.PasswordChangeSchema().bind(request=pyramid_csrf_request)
+    @pytest.fixture
+    def valid_data(self):
+        return {
+            "new_password": "valid_password",
+            "new_password_confirm": "valid_password",
+        }
+
+    @pytest.fixture
+    def schema(self, pyramid_csrf_request):
+        return schemas.PasswordAddSchema().bind(request=pyramid_csrf_request)
+
+
+@pytest.mark.usefixtures("user_password_service")
+class TestPasswordChangeSchema:
```

Comment:
`PasswordChangeSchema` hasn't changed in this PR (other than a small refactoring that doesn't affect behaviour) and the existing tests still passed without changes. But the existing tests weren't very good and didn't cover many cases, so I've rewritten them.